### DISCUSSION TITLE
Customizing json/dict serialization for custom field types

### DISCUSSION THREAD

**Discussion by nzig** [pydantic/pydantic#4243] [Pydantic V2]
I like the "Improvements to Dumping/Serialization/Export" section. I think "mode 3" would be especially useful.

I would like to propose an enhancement to the serialization mechanism, which would allow custom field types to specify how they should be converted to JSON-compliant types, similar to how `__pydantic_validation_schema__` would allow them to define how they should be validated.

The `Config.json_encoders` mechanism in the current pydantic is not as useful for this, because it requires that every model that includes the custom field type also includes its JSON encocder in its config. The V2 plan mentions
> kind of "3.5" mode where a python function is used to convert the data as the user wishes.

which I think would suffer from the same problem. Perhaps a serializng mechanism like this could be the best of both worlds:
1. If a conversion function is supplied call it on the field, and pass it to the next step
2. If the field type defines `__pydantic_serialize__`, call it and pass it to the next step
3. Apply pydantic's built-in serialization mechanism

I would love to hear your thoughts on this.

**Comment by samuelcolvin** [comment]
A method to look for on objects sounds fine, although it'll slow things down unless the approach is cached (you'd have to check every object an attribute). I guess we could make it configurable (a boolean as to whether or not to check for `__pydantic_validation_schema__`), but that's more settings, more confusion, more cognitive overhead.

Of course if we allow a function, it could easily look for `__pydantic_validation_schema__`.

In short, this area is TBD, I'm aware of your request, so will try to support it.

**Reply by nzig** [reply]
Thanks! 
You could decide to only look for the method on objects that are otherwise unserializable, and then there would be no performance hit (I think?). Similar to how subclasses of `str` won't have `__pydantic_validation_schema__` called even if they have it.

**Comment by j-carson** [comment]
I think I have a related problem: I have a Model with a field that is a small pandas dataframe.

This particular field needs to be serialized with `df.to_json` and deserialized with `pd.read_json` but there isn't a way to tell pydantic that.

**Reply by samuelcolvin** [reply]
That's not the same problem, you can use `json_loads` and `json_dumps` from [config](https://pydantic-docs.helpmanual.io/usage/model_config/)

**Reply by haoyun** [reply]
but `json_loads` and `json_dumps` are removed in `v2`...

**Reply by Samreay** [reply]
So I *think* the way to do this in Pydantic2 is using the annotations. Here's an example that solves a common issue where many systems send dates in messages as millisecond timestamps, but the same principle can be used for pandas dataframes @j-carson 

```python
from datetime import datetime
from datetime import timezone as tz

from pydantic import BaseModel
from pydantic.functional_serializers import PlainSerializer
from pydantic.functional_validators import PlainValidator

IntDatetime = Annotated[
    datetime,
    PlainValidator(lambda x: datetime.fromtimestamp(x / 1000, tz=tz.utc)),
    PlainSerializer(lambda x: int(x.timestamp() * 1000), when_used="always"),
]

class Test(BaseModel):
    start: IntDatetime
    normal_datetime: datetime


if __name__ == "__main__":
    now = datetime.now(tz=tz.utc)
    milliseconds_now = int(now.timestamp()) * 1000
    test = Test(start=milliseconds_now, normal_datetime=now)  # type: ignore
    print("Class is: ", test)
    print("Dump is:  ", test.model_dump_json())
```

And this outputs:
```bash
Class is:  start=datetime.datetime(2023, 9, 7, 4, 19, 48, tzinfo=datetime.timezone.utc) normal_datetime=datetime.datetime(2023, 9, 7, 4, 19, 48, 625569, tzinfo=datetime.timezone.utc)
Dump is:   {"start":1694060388000,"normal_datetime":"2023-09-07T04:19:48.625569Z"}
```

Please note that the ordering of the annotations is important. The validator must come before the serialiser, or the serialisation will not work on the model_dump_json

Though if anyone knows an annotation to add to make mypy realise I *can* pass an int into that annotated datetime, that would be great
